semgrep-core/src/engine/Match_extract_mode.ml

```diff
@@ -0,0 +1,363 @@
+                None
+            | Some (r, Some mval) -> Some (r, mval)
+            | None -> None))
+  (* Factor out rule *)
+  |> List.filter_map (function
+       | [] -> None
+       | (r, _) :: _ as xs -> Some (r, Common.map snd xs))
+  (* Convert mval match to offset of location in file *)
+  |> Common.map (fun (r, mvals) ->
+         ( r,
+           List.filter_map
+             (fun mval ->
+               let offsets = offsets_of_mval mval in
+               if Option.is_none offsets then report_no_source_range r;
+               offsets)
+             mvals ))
+  (* For every rule ... *)
+  |> Common.map (fun (r, offsets) ->
+         (* Sort matches by start position for merging *)
+         List.fast_sort (fun x y -> Int.compare x.start_pos y.start_pos) offsets
+         |> List.fold_left
+              (fun acc curr ->
+                match acc with
+                | [] -> [ curr ]
+                | last :: acc ->
+                    (* Keep both when disjoint *)
+                    if last.end_pos < curr.start_pos then curr :: last :: acc
+                      (* Filter out already contained range; start of
+                           last is before start of curr from sorting *)
+                    else if curr.end_pos <= last.end_pos then last :: acc
+                      (* Merge overlapping ranges *)
+                    else { last with end_pos = curr.end_pos } :: acc)
+              []
+         |> List.rev
+         (* Read the extracted text from the source file *)
+         |> Common.map (fun { start_pos; start_line; start_col; end_pos } ->
+                let source_file = open_in_bin xtarget.Xtarget.file in
+                let extract_size = end_pos - start_pos in
+                seek_in source_file start_pos;
+                let contents = really_input_string source_file extract_size in
+                logger#trace
+                  "Extract rule %s extracted the following from %s at bytes \
+                   %d-%d\n\
+                   %s"
+                  (fst r.Rule.id) xtarget.file start_pos end_pos contents;
+                (contents, map_loc start_pos start_line start_col xtarget.file))
+         (* Combine the extracted snippets *)
+         |> List.fold_left
+              (fun (consumed_loc, contents, map_contents) (snippet, map_snippet) ->
+                Buffer.add_string contents snippet;
+                let len = String.length snippet in
+                let snippet_lines, snippet_trailing =
+                  count_lines_and_trailing snippet
+                in
+                ( {
+                    consumed_loc with
+                    start_pos = consumed_loc.start_pos + len;
+                    start_line = consumed_loc.start_line + snippet_lines;
+                    start_col = snippet_trailing;
+                  },
+                  contents,
+                  (* Map results by chaining functions together, successively
```

Comment:
Would be helpful to note more clearly that this is how you're mapping back the results to the original. Also helpful to add a note that matches that cross boundaries are reported in the first snippet